gap/congruences/congsemigraph.gi

```diff
@@ -0,0 +1,271 @@
+                       ViewString(C!.H),
+                       ", W=",
+                       ViewString(C!.W),
+                       ">");
+end);
+
+InstallMethod(AsSemigroupCongruenceByGeneratingPairs,
+"for a congruence by Wang pair",
+[IsCongruenceByWangPair],
+function(cong)
+  local pairs, gens, verts, v, u, e, D, out, S, H, W;
+  S := Source(cong);
+  H := cong!.H;
+  W := cong!.W;
+  pairs := [];
+  gens := GeneratorsOfSemigroup(S);
+  verts := Filtered(gens, IsVertex);
+  D := GraphOfGraphInverseSemigroup(S);
+  out := OutNeighbours(D);
+  for v in H do
+    Add(pairs, [verts[v], MultiplicativeZero(S)]);
+  od;
+  for v in W do
+    for u in out[v] do
+      if not (u in H) then
+        for e in gens do
+          if Source(e) = verts[v] and Range(e) = verts[u] then
+            Add(pairs, [verts[v], e * e ^ -1]);
+          fi;
+        od;
+      fi;
+    od;
+  od;
+  return SemigroupCongruence(S, pairs);
+end);
+
+BindGlobal("SEMIGROUPS_MinimalHereditarySubsetsVertex",
+function(D, v)
+  local subsets, hereditary, u, out, s, a;
+  if IsMultiDigraph(D) then
+    ErrorNoReturn(D, " must not have multiple edges");
+  fi;
+  if not (v in DigraphVertices(D)) then
+    ErrorNoReturn(v, " must be a vertex of ", D);
+  fi;
+  out := Set(OutNeighboursMutableCopy(D)[v]);
+  subsets := [];
+  for u in [1 .. Length(out)] do
+    a := out[u];
+    RemoveSet(out, a);
+    hereditary := ShallowCopy(out);
+    for s in out do
+      UniteSet(hereditary, VerticesReachableFrom(D, s));
+    od;
+    if not (a in hereditary) and not (hereditary in subsets) then
+      AddSet(subsets, hereditary);
+    fi;
+    AddSet(out, a);
+  od;
+  return AsSortedList(subsets);
+end);
+
+InstallMethod(GISGeneratingCongruences,
+"for a graph inverse semigroup",
+[IsGraphInverseSemigroup],
+function(S)
+  local congs, out, v, h, D;
+  D := GraphOfGraphInverseSemigroup(S);
+  congs := [];
+  out := OutNeighbours(D);
+  for v in DigraphVertices(D) do
+    if Length(out[v]) = 0 then
+      Add(congs, CongruenceByWangPair(S, [v], []));
+    elif Length(out[v]) = 1 then
+      Add(congs, CongruenceByWangPair(S, [], [v]));
+    else
+      for h in SEMIGROUPS_MinimalHereditarySubsetsVertex(D, v) do
+        Add(congs, CongruenceByWangPair(S, h, [v]));
+      od;
+    fi;
+  od;
+  Add(congs, CongruenceByWangPair(S, [], []));
+  return congs;
+end);
+
+InstallMethod(AsCongruenceByWangPair, "for a semigroup congruence",
+[IsSemigroupCongruence],
+function(C)
+  local H, W, eq, j, Hgraph, Wgraph, v;
+  Hgraph := [];
+  Wgraph := [];
+  if not IsGraphInverseSemigroup(Source(C)) then
+    ErrorNoReturn(Source(C), " is not a graph inverse semigroup");
+  fi;
+
+  H := [];
+  W := [];
+  eq := EquivalenceRelationPartition(C);
+  eq := Filtered(eq, x -> ForAny(x, IsVertex));
+  for j in eq do
+    if MultiplicativeZero(Source(C)) in j then
+      Append(H, Filtered(j, IsVertex));
+    else
+      Append(W, Filtered(j, IsVertex));
```

Comment:
```suggestion
     UniteSet(W, List(Filtered(j, IsVertex), IndexOfVertexOfGraphInverseSemigroup));
```
Maybe `UniteSet` isn't the right thing to use here, can't remember, but basically can just union `W` and the other list.